sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -654,6 +654,25 @@ object DecorrelateInnerQuery extends PredicateHelper {
             val newProject = Project(newProjectList ++ referencesToAdd, newChild)
             (newProject, joinCond, outerReferenceMap)
 
+          case w@Window(projectList, partitionSpec, orderSpec, child) =>
+            val outerReferences = collectOuterReferences(w.expressions)
+            assert(outerReferences.isEmpty, s"Correlated column is not allowed in window " +
+              s"function: $w")
+            val newOuterReferences = parentOuterReferences ++ outerReferences
+            val (newChild, joinCond, outerReferenceMap) =
+              decorrelate(child, newOuterReferences, aggregated = true, underSetOp)
```

Comment:
window function is not an aggregate, indeed. Moreover, the query that you want to try, should give a runtime exception (As there are >1 rows returned from the subquery, unless only 1 row satisfies the b >= d filter).

However, what i want to achieve by setting aggregate = true is to avoid pulling-up non-equi predicates from under the Window function to the join above. It is correct to do so for equi-predicates, but for non-equi in general we need to keep a domain join. Perhaps the `aggregate` input param should be renamed to reflect that it is about non-equi predicate treatment..

Reply:
Yeah, I agree we need the logic of setting aggregated=true and that's fine. It's needed for correctness around outer ref substitution. And window funcs are doing something similar to aggregation, so I think it's ok as long as there's a comment.

You're right that this test case wouldn't work, how about the same thing but with a lateral join?

Reply:
we already have a test like that, see join-lateral.sql line 187:
https://github.com/apache/spark/pull/42383/files#diff-d8634e1fa795909252ff1b65501896fb5402befbad4ec9be86dfd840c5d3cfeaR187

Reply:
updated the comment